--- ../Tk804.024/Tixish/NoteBook.pm	Sat Jul 13 16:12:48 2002
+++ Tixish/NoteBook.pm	Sat Mar  1 00:53:38 2003
@@ -50,10 +50,10 @@
  my ($w, $args) = @_;

  $w->SUPER::Populate($args);
- $w->{'pad-x1'} = 0;
- $w->{'pad-x2'} = 0;
- $w->{'pad-y1'} = 0;
- $w->{'pad-y2'} = 0;
+ $w->{'pad-x1'} = undef;
+ $w->{'pad-x2'} = undef;
+ $w->{'pad-y1'} = undef;
+ $w->{'pad-y2'} = undef;

  $w->{'nWindows'} = 0;
  $w->{'minH'} = 1;
@@ -161,6 +161,10 @@
      my $myW = $w->Width;
      my $myH = $w->Height;

+     if (!defined $w->{'pad-x1'}) {
+	 $w->InitTabSize;
+     }
+
      my $cW = $myW - $w->{'pad-x1'} - $w->{'pad-x2'} - 2 * (defined $w->{-ipadx} ? $w->{-ipadx} : 0);
      my $cH = $myH - $w->{'pad-y1'} - $w->{'pad-y2'} - 2 * (defined $w->{-ipady} ? $w->{-ipady} : 0);
      my $cX = $w->{'pad-x1'} + (defined $w->{-ipadx} ? $w->{-ipadx} : 0);
@@ -378,13 +382,7 @@

     return unless Tk::Exists($w) && $w->{'nWindows'} && $w->{'resize'};

-    my ($tW, $tH) = $w->geometryinfo;
-    $w->{'pad-x1'} = 2;
-    $w->{'pad-x2'} = 2;
-    $w->{'pad-y1'} = $tH + (defined $w->{'-ipadx'} ? $w->{'-ipadx'} : 0) + 1;
-    $w->{'pad-y2'} = 2;
-    $w->{'minW'} = $tW;
-    $w->{'minH'} = $tH;
+    $w->InitTabSize;

     $w->{'resize'} = 0;
     my $reqW = $w->{-width} || 0;
@@ -434,6 +432,17 @@
     $w->{'counter'} = 0;
     $w->raise($w->{'topchild'} || ${$w->{'windows'}}[0]);
     $w->{'resize'} = 0;
+}
+
+sub InitTabSize {
+    my ($w) = @_;
+    my ($tW, $tH) = $w->geometryinfo;
+    $w->{'pad-x1'} = 2;
+    $w->{'pad-x2'} = 2;
+    $w->{'pad-y1'} = $tH + (defined $w->{'-ipadx'} ? $w->{'-ipadx'} : 0) + 1;
+    $w->{'pad-y2'} = 2;
+    $w->{'minW'} = $tW;
+    $w->{'minH'} = $tH;
 }

 1;
