--- Tk804.024/Tk/ErrorDialog.pm.orig	Mon Nov 11 11:11:53 2002
+++ Tk804.024/Tk/ErrorDialog.pm	Mon Nov 11 11:26:47 2002
@@ -3,54 +3,16 @@
 use vars qw($VERSION);
 $VERSION = '4.002'; # $Id: //depot/Tkutf8/Tk/ErrorDialog.pm#2 $
 
-use English;
 use Tk ();
 require Tk::Dialog;
 use base qw(Tk::Toplevel);
 
 
-# ErrorDialog - a translation of `tkerror' from Tcl/Tk to TkPerl.
+# ErrorDialog - a translation of bgerror() from Tcl/Tk to Perl/Tk.
 #
 # Currently TkPerl background errors are sent to stdout/stderr; use this
 # module if you want them in a window.  You can also "roll your own" by
 # supplying the routine Tk::Error.
-#
-# Stephen O. Lidie, Lehigh University Computing Center.  95/03/02
-# lusol@Lehigh.EDU
-#
-# This is an OO implementation of `tkerror', with a twist:  since there is
-# only one ErrorDialog object, you aren't required to invoke the constructor
-# to create it; it will be created automatically when the first background
-# error occurs.  However, in order to configure the ErrorDialog object you
-# must call the constructor and create it manually.
-#
-# The ErrorDialog object essentially consists of two subwidgets: an
-# ErrorDialog widget to display the background error and a Text widget for the
-# traceback information.  If required, you can invoke the configure() method
-# to change some characteristics of these subwidgets.
-#
-# Because an ErrorDialog object is a Frame widget all the composite base
-# class methods are available to you.
-#
-# Advertised widgets:  error_dialog, text.
-#
-# 1) Call the constructor to create the ErrorDialog object, which in turn
-#    returns a blessed reference to the new object:
-#
-#    require Tk::ErrorDialog;
-#
-#    $ED = $mw->ErrorDialog(
-#        -cleanupcode     => $code,
-#        -appendtraceback => $bool,
-#    );
-#
-#       mw -   a window reference, usually the result of a MainWindow->new
-#              call.
-#       code - a CODE reference if special post-background error processing
-#              is required (default is undefined).
-#       bool - a boolean indicating whether or not to append successive
-#              tracebacks (default is 1, do append).
-#
 
 use strict;
 
@@ -119,10 +81,10 @@
     $cw->ConfigSpecs(-cleanupcode => [PASSIVE => undef, undef, undef],
                      -appendtraceback => [ PASSIVE => undef, undef, 1 ]);
     $ED_OBJECT = $cw;
+    $cw->protocol('WM_DELETE_WINDOW' => sub {$cw->withdraw});
     return $cw;
 
-} # end new, ErrorDialog constructor
-
+} # end Populate
 
 sub Tk::Error {
 
@@ -159,6 +121,5 @@
     $w->break if ($ans =~ /skip/i);
 
 } # end Tk::Error
-
 
 1;
